### PR TITLE
[Model] Reduce MOSS TTS reference audio preparation overhead

### PR DESCRIPTION
**Pull request by Sy0307**
## Purpose

Cold MOSS reference encoding repeatedly converts Python waveform lists with `torch.tensor` and rebuilds the resampling filter for clips sharing the same sample-rate pair. Use a float32 NumPy array shared with PyTorch and a bounded, process-local LRU of 16 CPU resamplers. Explicit float32 filter construction preserves the existing functional resampler's samples.

This patch only changes MOSS reference preparation and its unit tests. Model encoding, batching, cache-hit behavior, scheduling, and chunk sizes remain unchanged. It is independent of MRv2.

## Test Plan

**vLLM Version:** 0.29.0; PyTorch 2.13.0, torchaudio 2.11.0, NumPy 2.3.5.

**vLLM-Omni Commit:** based on upstream main `d0755b571`.

CPU validation on the H200 host; no model weights or GPU execution required:

```bash
PYTHONPATH=. CUDA_VISIBLE_DEVICES= OMP_NUM_THREADS=1 python -m pytest -q \
  tests/model_executor/models/moss_tts/test_reference_encoder.py \
  -m 'core_model and cpu' --run-level core_model
```

The existing L1 CPU model-executor CI shard already collects this file. Tests cover mono/channel-first inputs, 16/24/32/44.1/48 kHz source rates, exact waveform equality with `functional.resample`, cache reuse/key separation, and no cross-clip filter state.

## Test Result

- **42 passed**, 14 deprecation warnings, 0.92s.
- Changed-file pre-commit hooks and whitespace checks passed.
- Current-main CPU microbenchmark: 20-second mono 48 kHz float32 waveform supplied as a Python list, converted/resampled to 24 kHz; one CPU thread, 5 warmups + 30 measurements per implementation. Median **61.67 ms → 35.66 ms**; ranges 60.48–77.84 / 35.56–35.84 ms. Prepared waveforms match exactly. This measures preparation only, not end-to-end serving.

### Earlier experimental serving evidence (not a benchmark of this main-based PR)

The same two changes were evaluated on a frozen experimental MOSS-TTS Local 1.5 MRv2 tree, vLLM 0.29, one H200 (both stages on the same card), CUDA Graphs, GPU reference encoder, single native `vllm bench serve` client. Dataset: CloudRipple ZH1000 / 133 references, **not original SeedTTS1088**; output cap512; codec chunks1→15. Fresh service/reference cache for each cold pass; hot pass repeats the full dataset. No speech warmup.

| Condition | Control TTFP p95 (ms) | Preparation patch TTFP p95 (ms) |
|---|---:|---:|
| C32 cold, two runs | 1543.49 / 1489.46 | 1085.37 / 1198.28 |
| C32 hot, two runs | 195.28 / 200.83 | 210.54 / 203.46 |
| C8 cold, one pair | 555.02 | 517.56 |
| C8 hot, one pair | 134.16 | 133.82 |

C32 cold mean of run-p95 values improved 24.7%, but C32 hot regressed 4.5%; the cause of that hot difference is unresolved. Throughput was approximately unchanged. **This is a cold-path candidate, not a demonstrated hot-latency improvement or a replacement for the accepted serving baseline.** All133 reference preparations were bitwise equal. Empty/near-output-cap generations remained in that experimental environment; ASR/speaker-quality equivalence was not evaluated. Current-main end-to-end serving remains unmeasured, hence this PR is a draft.

### CODE REVIEW

**Comment by hsliuustc0106** [comment]
This PR touches **vllm_omni/model_executor/, tests/model_executor/** (2 files). Based on CODEOWNERS coverage of the changed files, the most-related reviewers appear to be:

> @tzhouam @gcanlin @amy-why-3459

Could one of you take a look when you get a chance? Thanks!

File: vllm_omni/model_executor/models/moss_tts/reference_encoder.py [hunk 1]
```diff
@@ -46,15 +48,22 @@ def _sha1(s: str) -> str:
     return hashlib.sha1((s or "").encode("utf-8")).hexdigest()
 
 
+@lru_cache(maxsize=16)
+def _reference_resampler(sr: int, sr_target: int) -> torch.nn.Module:
```
Comment (BeatSeat) [inline vllm_omni/model_executor/models/moss_tts/reference_encoder.py:52]: Caching  instances globally via  without  and  in the key introduces state-mutation risks and device mismatches. If  is processed on GPU (e.g. in GPU reference encoder pipelines) or if any caller mutates the singleton via , subsequent executions will trigger . It is safer to cache the precomputed sinc filter weights  directly via  keyed by  rather than a mutable .

File: vllm_omni/model_executor/models/moss_tts/reference_encoder.py [hunk 2]
```diff
@@ -46,15 +48,22 @@ def _sha1(s: str) -> str:
...
+
+    # Match functional.resample's float32 kernel construction, rather than
+    # Resample's default float64 construction followed by a float32 cast.
+    return torchaudio.transforms.Resample(sr, sr_target, dtype=torch.float32)
+
+
 def _prep_wav_sync(wav_list: list, sr: int, sr_target: int) -> torch.Tensor:
     """Tensor-ise + resample one clip to ``sr_target`` (the blocking prep)."""
-    wav = torch.tensor(wav_list, dtype=torch.float32)
+    wav = torch.from_numpy(np.asarray(wav_list, dtype=np.float32))
```
Comment (BeatSeat) [inline vllm_omni/model_executor/models/moss_tts/reference_encoder.py:62]: (Non-blocking) Converting Python lists back via  is an ad-hoc fix for an upstream issue in , where the decoded NumPy array is converted into . Materializing hundreds of thousands of heap  objects only to repack them into NumPy and PyTorch tensors increases GC churn, which likely contributes to the observed 4.5% hot-path latency regression under C32. Retaining  across the serving interface in a follow-up would eliminate both conversion passes for all models.

File: tests/model_executor/models/moss_tts/test_reference_encoder.py [hunk 3]
```diff
@@ -534,3 +536,41 @@ async def test_encode_request_references_keys_second_speaker_by_slot(make_encode
...
+    import torchaudio
+
+    _reference_resampler.cache_clear()
+    try:
+        first = _reference_resampler(48000, _SR)
+        first(torch.ones(1, 4800))
+        assert _reference_resampler(48000, _SR) is first
+        assert _reference_resampler(16000, _SR) is not first
+        assert _reference_resampler(48000, 16000) is not first
+        silence = torch.zeros(1, 2400)
```
Comment (BeatSeat) [inline tests/model_executor/models/moss_tts/test_reference_encoder.py:572]: Testing for cross-clip state leakage using  creates a trivial assertion because linear convolution against zeros mathematically always returns zero. Any persistent buffer, state accumulator, or boundary padding pollution will be masked. We should pass a non-zero deterministic signal (e.g., random noise or a chirp) to genuinely verify that subsequent executions remain stateless.

File: vllm_omni/model_executor/models/moss_tts/reference_encoder.py (hunk 1 above)
Comment (Sy0307) [reply vllm_omni/model_executor/models/moss_tts/reference_encoder.py:55]: Current MOSS reference preparation always converts the NumPy/list input into a fresh CPU tensor at `_prep_wav_sync`, so the cached transform is only used with CPU inputs; there is no production `.to()` or mutation path. We also ran 32 concurrent H200 calls across multiple sample-rate keys with identical outputs; I am leaving this thread unresolved because the cache still stores a mutable `Resample` object, and a `(kernel, width)` cache can be considered separately if that stronger invariant is required.

File: vllm_omni/model_executor/models/moss_tts/reference_encoder.py (hunk 2 above)
Comment (Sy0307) [reply vllm_omni/model_executor/models/moss_tts/reference_encoder.py]: The rebased baseline now provides `_resolve_ref_audio_array`, and the MOSS non-Nano adapter uses it directly, so the production path covered by this PR no longer materializes a Python list before reference preparation. The broader list-free serving interface for other adapters is outside this PR; this outdated thread is resolved for the MOSS path.

File: tests/model_executor/models/moss_tts/test_reference_encoder.py (hunk 3 above)
Comment (Sy0307) [reply tests/model_executor/models/moss_tts/test_reference_encoder.py]: This test hunk comes from the rebased baseline and is no longer part of this PR’s diff; the PR’s added resampler test now drives `_prep_wav_sync` with a deterministic non-zero signal and verifies exact output plus cache hits. The original concern is therefore no longer applicable to this PR, so I am resolving the outdated thread.

**Comment by Sy0307** [comment]
Self-review: I rechecked the rebased head against upstream/main, the MOSS CPU reference-preparation path, cache ownership, numerical equivalence, and concurrent H200 behavior; the input-aliasing issue found during validation is fixed and the focused tests pass. The only remaining inline thread concerns the theoretical mutability of the private CPU-only Resample cache, so it remains unresolved for maintainer review rather than being marked complete.

File: vllm_omni/model_executor/models/moss_tts/reference_encoder.py (hunk 1 above)
Comment (Sy0307) [reply vllm_omni/model_executor/models/moss_tts/reference_encoder.py:55]: You are right that main already includes #7724: `_get_moss_processor()` moves the MOSS `audio_tokenizer` onto the code2wav-stage GPU, so my earlier wording about the whole reference encoder being CPU-only was too broad. The actual data flow still prepares a fresh CPU tensor in `_prep_wav_sync` before `encode_audios_from_wav` runs the GPU encoder; the device mismatch is therefore not reachable for this resampler, but the mutable cached-module concern remains and I am leaving the thread unresolved.

Comment (Sy0307) [reply vllm_omni/model_executor/models/moss_tts/reference_encoder.py:55]: Follow-up: main’s GPU reference encoder does not change this preprocessing boundary—`_prep_wav_sync` still performs CPU resampling before `audio_tokenizer` runs on the code2wav GPU. We will revisit device/dtype-keyed caching or an immutable kernel cache if resampling moves to GPU or this helper becomes a generic GPU-input API; the current PR intentionally keeps the CPU-only resampler cache.

**Review by gcanlin — APPROVED** [review]
